app/scripts/migrations/089.ts

```diff
@@ -0,0 +1,66 @@
+import { hasProperty, isObject } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+
+export const version = 89;
+
+/**
+ * Add an `id` to the `providerConfig` object.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(originalVersionedData: {
+  meta: { version: number };
+  data: Record<string, unknown>;
+}) {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  versionedData.data = transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, unknown>) {
+  if (
+    hasProperty(state, 'NetworkController') &&
+    isObject(state.NetworkController) &&
+    hasProperty(state.NetworkController, 'providerConfig') &&
+    isObject(state.NetworkController.providerConfig)
+  ) {
+    const { networkConfigurations, providerConfig } = state.NetworkController;
+
+    let newProviderConfigId;
+
+    if (!isObject(networkConfigurations)) {
+      return state;
+    }
+
+    for (const networkConfigurationId of Object.keys(networkConfigurations)) {
+      const networkConfiguration =
+        networkConfigurations[networkConfigurationId];
+      if (!isObject(networkConfiguration)) {
+        return state;
+      }
+      if (networkConfiguration.rpcUrl === providerConfig.rpcUrl) {
+        newProviderConfigId = networkConfiguration.id;
+      }
+    }
+
+    if (providerConfig.id || !newProviderConfigId) {
```

Comment:
There is still a branch here were no provider ID is found. In this case, the bug would still be present, wouldn't it? If so, this migration seems incomplete.

Built-in networks have no id, so it'll still be missing of one of those is selected. I guess it's also possible that someone has a selected network that doesn't match any configured network (though that isn't _supposed_ to be possible), which we could handle by discarding the invalid current selected network config.

Reply:
@Gudahtt Built in networks don't hit this bug, because they are selected by comparing the `type` on the provider config to the network config

Reply:
Though it seems fine to leave that case unhandled instead if you'd prefer we deal with it later

Reply:
I will leave this case unhandled for now. In addition to wanting to merge this PR as soon as possible: what if users don't want to be redirected to the default? e.g we'd have to handle the case of users that set a custom rpc during onboarding. I'd rather avoid those concerns for now, and users can surface these other edge cases if they exist